ACCUMULO-4500 ACCUMULO-96 Added summarization

keith-turner opened the pull request.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/admin/NewTableConfiguration.java:
```diff
@@ -42,7 +46,13 @@
+  private Map<String,String> samplerProps = Collections.emptyMap();
+  private Map<String,String> summarizerProps = Collections.emptyMap();
+
+  private void checkDisjoint(Map<String,String> props, Map<String,String> sampleProps, Map<String,String> summrizerProps) {
```
s/summrizerProps/summarizerProps/

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/admin/NewTableConfiguration.java:
```diff
@@ -121,8 +128,22 @@ public NewTableConfiguration setProperties(Map<String,String> prop) {
-    this.samplerConfiguration = samplerConfiguration;
+    Map<String,String> tmp = new SamplerConfigurationImpl(samplerConfiguration).toTablePropertiesMap();
+    checkDisjoint(properties, tmp, summarizerProps);
+    this.samplerProps = tmp;
```
Seems odd to be checking disjoint with `summarizerProps` but then updating `samplerProps` instead. Maybe javadoc on those member variables would help describe what they each hold.

keith-turner replied:
Checking summarizerProps is completely unnecessary here. The function is used in three places.  Only in one of those three places does it actually need to check both.

keith-turner replied:
I cleaned this up and simplified it in commit 40ed96f.  That removed the unnecessary checks and hopefully made it a little more clear.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/admin/NewTableConfiguration.java:
```diff
@@ -121,8 +128,22 @@ public NewTableConfiguration setProperties(Map<String,String> prop) {
     requireNonNull(samplerConfiguration);
-    SamplerConfigurationImpl.checkDisjoint(properties, samplerConfiguration);
-    this.samplerConfiguration = samplerConfiguration;
+    Map<String,String> tmp = new SamplerConfigurationImpl(samplerConfiguration).toTablePropertiesMap();
```
Maybe a static utility to see if we can do an (unsafe) cast to `SamplerConfigurationImpl` to avoid an unnecessary object creation?

keith-turner replied:
I don't think this would be called frequently enough for it to be worth it.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java:
```diff
@@ -0,0 +1,107 @@
+public interface SummaryRetriever {
+
+  /**
+   * Summary data is only retrieved from data that has been written to files. Data recently written to Accumulo may be in memory and there will not show up in
```
Using 3-person declarative (the style used on the other methods) is good for Javadoc, IMO.

e.g. `Forces a flush of data in tablets before summary data is retrieved. ....:

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java:
```diff
@@ -0,0 +1,107 @@
+  SummaryRetriever flush(boolean shouldFlush);
+
+  /**
+   * Allows optionally setting start row before retrieving data. The start row is not inclusive.
```
Nit: Move "optionally" to it's own sentence (e.g. "The start row is not inclusive. Calling this method is optional.") or lead with it (e.g. "Optional, allows setting a start row").

Same applies to the other methods.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java:
```diff
@@ -0,0 +1,107 @@
+
+  /**
+   * Gets summaries generated with a configuration that matches the given regex. For a given SummarizationConfiguration it is matched in exactly the following
+   * way. This allows the regex to match on classname and options.
```
It's not obvious to me what this method is doing under the hood and I don't think the code snippet helps.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:
```diff
@@ -808,4 +812,64 @@ void setSamplerConfiguration(String tableName, SamplerConfiguration samplerConfi
+   * This is a entry point for retrieving summaries with optional restrictions.
+   *
+   * <p>
+   * Inorder to retrieve Summaries, the Accumulo user making the request will need the {@link TablePermission#GET_SUMMARIES} table permission.
```
s/Inorder/In order/

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:
```diff
@@ -808,4 +812,64 @@ void setSamplerConfiguration(String tableName, SamplerConfiguration samplerConfi
+   * Since summary data is cached, its important to use the summary selection options to only read the needed data into the cache.
+   *
+   * <p>
+   * Summary data will be merged on the tablet servers and then in this client process. Therefore its important that the required summarizers are on the clients
```
s/its/it's/

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/impl/TableOperationsImpl.java:
```diff
@@ -1661,4 +1676,138 @@ public Locations locate(String tableName, Collection<Range> ranges) throws Accum
+  @Override
+  public SummaryRetriever getSummaries(String tableName) {
+
+    return new SummaryRetriever() {
```
Pull this out to a private static class instead of doing it inline? Would help testing.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/impl/TableOperationsImpl.java:
```diff
@@ -1661,4 +1676,138 @@ public Locations locate(String tableName, Collection<Range> ranges) throws Accum
+        ClientContext cct = new ClientContext(context.getInstance(), context.getCredentials(), context.getConfiguration()) {
+          @Override
+          public long getClientTimeoutInMillis() {
+            return Math.max(super.getClientTimeoutInMillis(), 60 * 60 * 1000);
```
This is forcing a minimum client timeout of 60mins which could be irritating for someone who set a 10minute timeout. Is this what you intended?

keith-turner replied:
Ugh.  Glad you found this.  I meant to circle back to this, but forgot about it.  When I did this hack I Was worried about the case where getting summaries is slow (because none are in cache) and a result the client times and initiates a new operation (while threads for old operation are still churning).  I need a better solution.

keith-turner replied:
I switched this to a timeout of zero, which is what some other long RPC ops currently do.

joshelser replied:
Correct me if I'm wrong, but don't we usually do that for client-facing RPCs? As-in, while the client is still there (connected), we let the RPC sit. My worry is that if we do this inside the TabletServer, the risk is larger since that is a Java process which we expect to run for days/months/indefinitely.

Precedence to the contrary is fine, as well.

keith-turner replied:
> Correct me if I'm wrong, but don't we usually do that for client-facing RPCs?

That is a good point.  I think so.

> My worry is that if we do this inside the TabletServer, the risk is larger since that is a Java process which we expect to run for days/months/indefinitely.

One thing I am unsure about is the relationship between the timeout and TCP keepalive.  From the research I have done this morning and previously, I think the timeout works mechanism above the socket and just closes it if the operation takes longer than expected (even if the remote machine is fine and working).  In this case I would rather wait indefinitely and rely on something like TCP keep alive mechanism to close the socket if the remote machine is unresponsive.  I am going to do an experiment where I kill a tserver that I am waiting on with no timeout and see what happens (if something is enabling tcp keepalive, the socket should still be closed eventually causing a retry).

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/impl/TableOperationsImpl.java:
```diff
@@ -1661,4 +1676,138 @@ public Locations locate(String tableName, Collection<Range> ranges) throws Accum
+  @Override
+  public void addSummarizers(String tableName, SummarizerConfiguration... newConfigs) throws AccumuloException, AccumuloSecurityException,
+      TableNotFoundException {
+    HashSet<SummarizerConfiguration> currentConfigs = new HashSet<>(SummarizerConfiguration.fromTableProperties(getProperties(tableName)));
```
`Collections.singleton` instead of `new HashSet<>`?

keith-turner replied:
The function returns a collection, which could have multiple.

joshelser replied:
Ahh, I missed that. I just expected this method to return a single SummarizeConfiguration. Makes sense.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/impl/TableOperationsImpl.java:
```diff
@@ -1661,4 +1676,138 @@ public Locations locate(String tableName, Collection<Range> ranges) throws Accum
+
+    newConfigSet.removeIf(sc -> currentConfigs.contains(sc));
+
+    Set<String> newIds = newConfigSet.stream().map(sc -> sc.getPropertyId()).collect(toSet());
```
Such wizardy.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/mapred/AccumuloFileOutputFormat.java:
```diff
@@ -136,6 +138,20 @@ public static void setSampler(JobConf job, SamplerConfiguration samplerConfig) {
   }
 
+  /**
+   * Specify a list of summarizer configurations to create summary data in the output file. Each Key Value written will be passed to the configured
```
s/Specify/Specifies/

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/mapreduce/AccumuloFileOutputFormat.java:
```diff
@@ -134,6 +136,20 @@ public static void setSampler(Job job, SamplerConfiguration samplerConfig) {
   }
 
+  /**
+   * Specify a list of summarizer configurations to create summary data in the output file. Each Key Value written will be passed to the configured
```
s/Specify/Specifies/

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/mapreduce/lib/impl/FileOutputConfigurator.java:
```diff
@@ -209,4 +210,12 @@ public static void setSampler(Class<?> implementingClass, Configuration conf, Sa
+    Map<String,String> props = SummarizerConfiguration.toTableProperties(sumarizerConfigs);
+
+    for (Entry<String,String> entry : props.entrySet()) {
+      conf.set(enumToConfKey(implementingClass, Opts.ACCUMULO_PROPERTIES) + "." + entry.getKey(), entry.getValue());
```
Use a `StringBuilder` to cut down on transient `String`s you're generating.

keith-turner replied:
I don't think this code is called frequently enough to justify optimizations that would make the code longer/harder to read.  Also the java compiler will use StringBuilder when concatenating  multiple strings.  However, I don't think it will reuse it across the loop.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/rfile/RFileWriterBuilder.java:
```diff
@@ -59,15 +62,22 @@ OutputStream getOutputStream() {
+  private Map<String,String> samplerProps = Collections.emptyMap();
+  private Map<String,String> summarizerProps = Collections.emptyMap();
+
+  private void checkDisjoint(Map<String,String> props, Map<String,String> sampleProps, Map<String,String> summrizerProps) {
```
Same method you have in NewTableConfiguration. Consolidate?

keith-turner replied:
That would be the right thing to do.  Its a bit annoying because its private in classes that in the public API space.  So I would need to place in some third class that not in a public API package, and I don't know where to put it.  I will added a comment to both methods mentioning its doppelganger .

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/rfile/RFileWriterBuilder.java:
```diff
@@ -92,10 +102,11 @@ public RFileWriter build() throws IOException {
     } else {
       return new RFileWriter(fileops.newWriterBuilder().forFile(out.path.toString(), out.getFileSystem(), out.getConf()).withTableConfiguration(acuconf)
-          .build(), visCacheSize);
+          .setAccumuloStartEnabled(false).build(), visCacheSize);
```
Why `false`?

keith-turner replied:
It presumed that this code us running in a client process, so do not want to use Accumulo start to load classes.

joshelser commented on core/src/main/java/org/apache/accumulo/core/file/FileOperations.java:
```diff
@@ -301,6 +302,16 @@ public RateLimiter getRateLimiter() {
+    }
+
+    @Override
+    public OpenWriterOperation setAccumuloStartEnabled(boolean enableAccumuloStart) {
```
This is really vague to me. I don't know what this actually means..

keith-turner replied:
I added documentation to the method declaration in the super type.

keith-turner replied:
The one thing I miss from Review Board is its per review issues.  Those were nice to help track what needed to be done on a review with lots of comments.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/summary/CountingSummarizer.java:
```diff
@@ -0,0 +1,302 @@
+   */
+  public static final String SEEN_STAT = "seen";
+
+  // this default can not be changed as persisted summary data depends on it
```
Can you expand on this comment?

keith-turner replied:
I should add some documentation about considerations for persistence to the super type.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/summary/CountingSummarizer.java:
```diff
@@ -0,0 +1,302 @@
+      private UnaryOperator<K> copier = copier();
+
+      private void incrementCounter(K counter) {
+        emitted++;
```
Do we care about over-counting `emitted`? e.g. what happens when `encoded.apply(counter)` would throw an exception?

keith-turner replied:
Currently it would bork the compaction, so the counter issue would be moot.

I did think about making behavior for failures configurable.  If a summarizer fails, we could record that fact that it ran and failed in the rfile OR let it bork the compaction ... depending on config.

joshelser replied:
> Currently it would bork the compaction, so the counter issue would be moot.

SGTM can deal with your other thoughts as a follow-on improvement.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/summary/Summary.java:
```diff
@@ -0,0 +1,126 @@
+ * from.
+ *
+ * @see Summarizer
+ */
```
needs since tag

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/summary/CountingSummarizer.java:
```diff
@@ -0,0 +1,302 @@
+  @Override
+  public Combiner combiner(SummarizerConfiguration sc) {
+    init(sc);
+    return new Combiner() {
```
Make this a private static class instead of inline to make testing `merge` easier

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/summary/summarizers/FamilySummarizer.java:
```diff
@@ -0,0 +1,36 @@
+import org.apache.accumulo.core.data.ArrayByteSequence;
+import org.apache.accumulo.core.data.ByteSequence;
+
+public class FamilySummarizer extends CountingSummarizer<ByteSequence> {
```
needs javadoc & since tag

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/summary/summarizers/VisibilitySummarizer.java:
```diff
@@ -0,0 +1,37 @@
+import org.apache.accumulo.core.data.ArrayByteSequence;
+import org.apache.accumulo.core.data.ByteSequence;
+
+public class VisibilitySummarizer extends CountingSummarizer<ByteSequence> {
```
needs javadoc & since tag

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/summary/SummarizerConfiguration.java:
```diff
@@ -0,0 +1,238 @@
+      return this;
+    }
+
+    public Builder addOption(String key, String value) {
```
These methods need javadoc.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/summary/SummarizerConfiguration.java:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    /**
+     * Setting this is optional. If not set, an id is generated using hashing that will likely be unique.
```
3rd-person declarative, please.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/summary/SummarizerConfiguration.java:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    public Builder addOption(String key, long value) {
+      return addOption(key, value + "");
```
`Integer.toString(value)`

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/summary/SummarizerConfiguration.java:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    public Builder addOptions(String... keyValuePairs) {
+      Preconditions.checkArgument(keyValuePairs.length % 2 == 0, "Require an even number of arguments, got %s", keyValuePairs.length);
```
Check for positive numbers too?

mikewalch commented on docs/src/main/asciidoc/chapters/summaries.txt:
```diff
@@ -0,0 +1,209 @@
+dedicated cache for summary data on each tserver with a configurable size.  In
+order for summary data to fit in cache, it should probably be small.
+
+For information on writing a custom summarizer see the javadoc for
```
Should mention that there are summarizers available for use (and list them) in addition to creating your own custom serializer

keith-turner replied:
I added a sentence mentioning what package to look in for summarizers that ship with Accumulo.  I really really with there was a good way to link to javadocs from the user manual.  Could create much better documentation.

keith-turner commented on docs/src/main/asciidoc/chapters/summaries.txt:
```diff
@@ -0,0 +1,209 @@
++org.apache.accumulo.core.client.summary.Summarizer+.
+
+=== Configuring
+
```
In the rendered ascii doc this will link the prefix.  The prefix property has documentation on how to configure a summarize which will show up in the ascii doc. Did you see that documentation?

keith-turner commented:
TODO now that there is summary cache, need to include it in the tservers memory size sanity checks.

mjwall commented:
@keith-turner we talked about this yesterday, but I wanted to post it here.  What would happen if a file is deleted, like maybe compacted and gc'd, after the file list is grabbed?

keith-turner commented:
> @keith-turner we talked about this yesterday, but I wanted to post it here. What would happen if a file is deleted, like maybe compacted and gc'd, after the file list is grabbed?

@mjwall I had not thought of this case and currently have no handling for it.  Yet another win for code reviews.

I think the best solution to this problem is to introduce a new inaccuracy counter called `deleted`.  There are already a few inaccuracy counters reported when gather summary information.  I will add another comment that shows where these can be found.

At first I thought I could circle back and use the file that replaced a missing file.  However this approach has a problem.  Multiple deleted files could have been compacted into the replacement file, and for some of those deleted files we may have already gathered and merged summary information.  Trying to avoid this problem would make gathering summaries more expensive.  In order to keep gathering summaries fast, I think it would be best to just report the problem.  If someone really wants to avoid this problem, they can clone the table and make the request against the clone.  I can put this avoidance strategy in the javadoc for `deleted`.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/summary/Summary.java:
```diff
@@ -0,0 +1,127 @@
+ */
+public class Summary {
+
+  public static class FileStatistics {
```
@mjwall I was thinking of adding the `deleted` statistic to this class.

joshelser commented on docs/src/main/asciidoc/chapters/summaries.txt:
```diff
@@ -0,0 +1,211 @@
+ root@uno summary_test> summaries
+  Summarizer         : org.apache.accumulo.core.client.summary.summarizers.VisibilitySummarizer vis {}
+  File Statistics    : [total:1, missing:0, extra:0, large:0]
+  Summary Statistics : 
```
So cool.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/conf/PropertyType.java:
```diff
@@ -117,6 +121,7 @@ String getFormatDescription() {
    * @return true if value is valid or null, or if this type has no regex
    */
   public boolean isValidFormat(String value) {
+    Preconditions.checkState(predicate != null, "Predicate was null, maybe this enum was serialized????");
```
Enums are inherently java-serializable. This cannot be prevented. What this is trying to work around (in conjunction with the predicate being transient) is that the predicates (especially the lambdas?) are not themselves serializable and cannot necessarily be included with the enum. If, for whatever reason, these enums ever get serialized, and then deserialized, we won't necessarily be able to call the `isValidFormat` function on them.

I'm not sure if this can even ever happen. I would expect enums to have one, and only one, instance in the JVM ever.... the one created by the constructor... not the one produced by deserializing. So, this might not ever be able to happen... but it'd be interesting to see if it did.

keith-turner replied:
For some reason the changes I made resulted in findbugs complaining about code I didn't change.  These changes were a result of a findbugs complaint.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/summary/SummarySerializer.java:
```diff
@@ -0,0 +1,543 @@
+        lg.save(dos, symbolTable);
+      }
+
+      dos.close();
```
try-with-resources even better.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/summary/SummarySerializer.java:
```diff
@@ -0,0 +1,543 @@
+      }
+
+      int end = summaries.size();
+      if (end % 2 == 1) {
```
There lower bound is determined by the method that calls it.  That method keeps calling collapse as long as its greater than a certain size.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/summary/SummarySerializer.java:
```diff
@@ -0,0 +1,543 @@
+    private void saveSummary(DataOutputStream dos, HashMap<String,Integer> symbolTable, Map<String,Long> summary) throws IOException {
+      WritableUtils.writeVInt(dos, summary.size());
+      for (Entry<String,Long> e : summary.entrySet()) {
+        WritableUtils.writeVInt(dos, symbolTable.get(e.getKey()));
```
It should always have a mapping.  It it does not the code will fail fast when trying to convert `null` to `int`.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/summary/SummaryInfo.java:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  Text getLastRow() {
+    return lastRow;
```
fixed this

keith-turner commented on core/src/main/java/org/apache/accumulo/core/summary/SummaryCollection.java:
```diff
@@ -0,0 +1,167 @@
+    }
+  }
+
+  public void merge(SummaryCollection other, SummarizerFactory factory) {
```
I don't think so.  Alot of code is assumed to be single threaded.  Also, its internal as you mentioned.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/summary/summarizers/DeletesSummarizer.java:
```diff
@@ -0,0 +1,75 @@
+
+  @Override
+  public Combiner combiner(SummarizerConfiguration sc) {
+    return (m1, m2) -> m2.forEach((k, v) -> m1.merge(k, v, Long::sum));
```
I am not sure about this.  The code is already so short.  But if I did do it, `LongSummarizer` is too generic of a name because it summing.   So maybe it could be called `SummingSummarizer`.

joshelser replied:
My guess is that summations would be the most common summaries that we'd want to do in Accumulo so, even if the boilerplate reduction is small for each, there might be some long-term gain for all of the uses of summation-summaries.

joshelser commented on test/src/main/java/org/apache/accumulo/test/functional/SummaryIT.java:
```diff
@@ -0,0 +1,759 @@
+    Assert.assertEquals(1, summary.getFileStatistics().getTotal());
+    Assert.assertEquals(1, summary.getFileStatistics().getExtra());
+    long total = summary.getStatistics().get("total");
+    Assert.assertTrue(total > 0 && total <= 10_000);
```
Nit: list what the total was in the message argument for `assertTrue(String, boolean)`.

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:
```diff
@@ -808,4 +812,64 @@ void setSamplerConfiguration(String tableName, SamplerConfiguration samplerConfi
+   * @since 2.0.0
+   * @see Summarizer
+   */
+  SummaryRetriever getSummaries(String tableName) throws TableNotFoundException, AccumuloException, AccumuloSecurityException;
```
Is there a better method name than `getSummaries`? I didn't notice it was strange until I got to the tests where you have (paraphrasing)

`conn.tableOperations().getSummaries("my_table").retrieve()`

The `SummaryRetriever` is more like a builder, right?

keith-turner replied:
Yeah the current name does not communicate well what it is.  I really don't like the current name.  I tried to think of something better and got sidetracked.

keith-turner replied:
I renamed it to `summarize()`  still not sure about that name, but I like it better than `getSummaries()`

joshelser commented on test/src/main/java/org/apache/accumulo/test/functional/SummaryIT.java:
```diff
@@ -0,0 +1,759 @@
+    }
+  }
+
+  public static class FooCS extends CompactionStrategy {
```
Can you provide a better name and/or javadoc?

joshelser commented on test/src/main/java/org/apache/accumulo/test/functional/SummaryIT.java:
```diff
@@ -0,0 +1,759 @@
+    }
+  }
+
+  public static class FooFilter extends Filter {
```
Ditto on name/javadoc.

joshelser commented on test/src/main/java/org/apache/accumulo/test/functional/SummaryIT.java:
```diff
@@ -0,0 +1,759 @@
+    } catch (PatternSyntaxException e) {}
+  }
+
+  public static class FooCounter implements Summarizer {
```
Ditto name/javadoc.

joshelser commented on test/src/main/java/org/apache/accumulo/test/functional/BasicSummarizer.java:
```diff
@@ -0,0 +1,80 @@
+
+  public static final String DELETES_STAT = "deletes";
+  public static final String MIN_STAMP_STAT = "minStamp";
+  public static final String MAX_STAMP_STAT = "maxStamp";
```
s/STAMP/TIMESTAMP/ ?

joshelser commented on test/src/main/java/org/apache/accumulo/test/ShellServerIT.java:
```diff
@@ -1894,4 +1896,71 @@ private String getTableId(String tableName) throws Exception {
+    ts.exec("createtable summary");
+    ts.exec("config -t summary -s table.summarizer.del=" + DeletesSummarizer.class.getName());
+    ts.exec("config -t summary -s table.summarizer.fam=" + FamilySummarizer.class.getName());
+    // ts.exec("config -t summary -s table.summarizer.fam.opt."+CountingSummarizer.INGNORE_DELETES_OPT+"=false");
```
nit: delete

joshelser commented on test/src/main/java/org/apache/accumulo/test/ShellServerIT.java:
```diff
@@ -1894,4 +1896,71 @@ private String getTableId(String tableName) throws Exception {
+    ts.exec("flush -t summary -w");
+
+    String output = ts.exec("summaries");
+    Assert.assertTrue(output.matches("(?sm).*^.*deletes\\s+=\\s+0.*$.*"));
```
Please provide some debugging here so that we would know what the output was if the assertion ever fails.

keith-turner replied:
In 7a9fa75 I created a new utility method called `assertMatches` that prints the regex and output on failure.  A nice improvement.

joshelser commented on server/tserver/src/main/java/org/apache/accumulo/tserver/compaction/strategies/TooManyDeletesCompactionStrategy.java:
```diff
@@ -0,0 +1,173 @@
+ *
+ * @since 2.0.0
+ */
+public class TooManyDeletesCompactionStrategy extends DefaultCompactionStrategy {
```
I can't adequately express how cool I think this is.

joshelser commented on server/tserver/src/main/java/org/apache/accumulo/tserver/compaction/strategies/ConfigurableCompactionStrategy.java:
```diff
@@ -48,6 +53,83 @@ void gatherInformation(MajorCompactionRequest request) {}
+          }
+
+          if (!needsCompaction) {
+            okFiles.add(fref);
```
It looks like `okFiles` is only set once (to `Collections.emptyMap()`). Isn't the `Map` returned by that method immutable?

keith-turner replied:
Yikes!  Thats a really nice find.

keith-turner replied:
I Added an IT for these shell options in  	7a9fa75

joshelser commented on core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java:
```diff
@@ -0,0 +1,513 @@
+    if (summarizerPattern != null) {
+      Pattern pattern = Pattern.compile(summarizerPattern);
+      // The way conf is converted to string below is documented in the API, so consider this when making changes!
+      summarySelector = conf -> pattern.matcher(conf.getClassName() + " " + new TreeMap<>(conf.getOptions())).matches();
```
Oh my...

joshelser commented on core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java:
```diff
@@ -0,0 +1,513 @@
+
+          TSummaries tSums = null;
+          try {
+            tSums = client.getSummariesFromFiles(tinfo, ctx.rpcCreds(), getRequest(), files);
```
Should we have some kind of inspection as to this cross-sectional RPC? What kind of metrics/logging can we add to understand this process better?

If I want to get the summaries for a table as a client, and that table is hosted by every tserver, so that mean we might get a cartesian product of RPCs from tserver->tserver?

joshelser commented on core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java:
```diff
@@ -0,0 +1,513 @@
+          try {
+            tSums = client.getSummariesFromFiles(tinfo, ctx.rpcCreds(), getRequest(), files);
+          } catch (TApplicationException tae) {
+            throw tae;
```
What about the case of a failing tserver? Wouldn't this propagate up and fail the entire getSummaries call? Do we want to try a couple of times to get the summaries from the tserver? Maybe try a diffferent tserver?

keith-turner replied:
Yes, I think it would fail the entire call and that is the intent.  The client should get an `AccumuloServerException`

All other accumulo code fails when an unexpected server side exception is seen. I would not want to do something different here. If anything different was done, would want to consider doing it for all code.

joshelser commented on core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java:
```diff
@@ -0,0 +1,513 @@
+          HostAndPort location = HostAndPort.fromString(entry.getKey());
+          Map<String,List<TRowRange>> allFiles = entry.getValue();
+
+          Future<ProcessedFiles> future = execSrv.submit(new FilesProcessor(tinfo, location, allFiles));
```
From `Executors.newCachedThreadPool()` javadoc:

> These pools will typically improve the performance of programs that execute many short-lived asynchronous tasks

But, in `FilesProcessor` aren't we running an RPC without timeout? Seem to be competing goals (and potentially keep us to hold on to a lot of threads if RPCs are delayed).

joshelser commented on core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java:
```diff
@@ -0,0 +1,513 @@
+          filesGBL = getFilesGroupedByLocation(fileSelector);
+          UtilWaitThread.sleep(250);
+        } else {
+          break;
```
Just make this `return partitionSummaries;`? Not sure if I'm missing an execution branch.

joshelser commented on core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java:
```diff
@@ -0,0 +1,513 @@
+      try {
+        fileSummaries.merge(future.get(), factory);
+      } catch (ExecutionException | InterruptedException e) {
+        throw new RuntimeException(e);
```
Should be a good citizen and re-set the interrupted state when catching `InterruptedExeption`.

joshelser commented on core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java:
```diff
@@ -0,0 +1,513 @@
+
+  public SummaryCollection gather() throws TableNotFoundException, AccumuloException, AccumuloSecurityException {
+
+    ExecutorService es = Executors.newFixedThreadPool(ctx.getConnector().instanceOperations().getTabletServers().size());
```
More surprise threads??

keith-turner replied:
Yeah this not ideal.  These threads will not consume CPU, they are just waiting on remote operations.  The threads will consume memory resources though.  One nice solution for this would be to make all of the request asynchronously AND not use threads.  I looked into doing this, it would be a lot of new code that would need to be tested for fault tolerance.  I will open a follow up issue for this.

joshelser replied:
Sounds good.

joshelser commented on core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java:
```diff
@@ -0,0 +1,513 @@
+        try {
+          allSummaries.merge(future.get(), factory);
+        } catch (InterruptedException | ExecutionException e) {
+          throw new AccumuloException(e);
```
re-set interrupted state?

keith-turner commented:
@joshelser I have been thinking about the issues you brought up regarding the thread pools and RPC timeouts.  These issues will cause problems when there are problems on the cluster (and that never happens :).  I am going to work on fixing them.  Thinking of introducing configurable thread pools as needed.  Also going to make RPC work like scanner with sessions and sessions timeouts.

joshelser commented:
>  These issues will cause problems when there are problems on the cluster (and that never happens :)

It's so convenient that we always have perfect computers, networks, and administrators :)

>  I am going to work on fixing them. Thinking of introducing configurable thread pools as needed. Also going to make RPC work like scanner with sessions and sessions timeouts.

LMK if you have ideas you want to bounce around. I was a bit brutal in the first review -- we kind of live in a world right now where our RPC code can just let stuff run for a very long time (you inherited a bit of the problem). I don't want to completely block this feature, but I also want to make sure we don't make the problem worse.

keith-turner commented:
> I was a bit brutal in the first review -- we kind of live in a world right now where our RPC code can just let stuff run for a very long time (you inherited a bit of the problem).

There were so many things to consider for this feature.  I am really glad you made me focus on this aspect.  

>  we kind of live in a world right now where our RPC code can just let stuff run for a very long time

The session concept that some of the code currently uses (like scans) defends against the following : 

 * Doing work for a client that is no longer there.
 * Clients starting a second instance of a task, while the first is still processing.

The current code is susceptible to both of the problems above and I would like to avoid that.

keith-turner commented:
I think I have addressed all of the code review issues, except for metrics/tracing.  I think I will open a follow on issue to instrument this feature.  The thread pool and rpc timeout issues are no more.

joshelser commented on core/src/main/java/org/apache/accumulo/core/conf/Property.java:
```diff
@@ -362,8 +362,14 @@
   TSERV_SLOW_FLUSH_MILLIS("tserver.slow.flush.time", "100ms", PropertyType.TIMEDURATION,
       "If a flush to the write-ahead log takes longer than this period of time, debugging information will written, and may result in a log rollover."),
+  TSERV_SUMMARY_PARTITION_THREADS("tserver.summary.partition.threads", "10", PropertyType.COUNT,
+      "Summary data must be retreied from files.  For a large number of files, the files are broken into partitions of 100K files.  This setting determines "
```
s/retreied/retrieved/

joshelser commented on server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java:
```diff
@@ -1812,7 +1884,26 @@ public TSummaries getSummariesFromFiles(TInfo tinfo, TCredentials credentials, T
+    public TSummaries contiuneGetSummaries(TInfo tinfo, long sessionId) throws NoSuchScanIDException, TException {
+      SummarySession session = (SummarySession) sessionManager.getSession(sessionId);
+      if (session == null) {
+        throw new NoSuchScanIDException();
```
Is it OK to be re-using the same exception as we use in the Scanner path?

keith-turner replied:
That exception is used in a few other places (that are not scan) to indicate no such session id.  I was just following that.  The exception should be named no such session exception.

joshelser commented on server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServerResourceManager.java:
```diff
@@ -232,7 +234,9 @@ public TabletServerResourceManager(TabletServer tserver, VolumeManager fs) {
-    summaryRetrievalPool = createIdlingEs(Property.TSERV_SUMMARY_RETRIEVAL_THREADS, "summary retriever", 60, TimeUnit.SECONDS);
+    summaryRetrievalPool = createIdlingEs(Property.TSERV_SUMMARY_RETRIEVAL_THREADS, "summary file retriever", 60, TimeUnit.SECONDS);
+    summaryRemotePool = createIdlingEs(Property.TSERV_SUMMARY_REMOTE_THREADS, "summary remote", 60, TimeUnit.SECONDS);
+    summaryParitionPool = createIdlingEs(Property.TSERV_SUMMARY_PARTITION_THREADS, "summary partition", 60, TimeUnit.SECONDS);
```
s/summaryParitionPool/summaryPartitionPool/

joshelser commented on test/src/main/java/org/apache/accumulo/test/functional/SummaryIT.java:
```diff
@@ -770,4 +774,44 @@ public void testExceptions() throws Exception {
+    int q = 0;
+
+    SortedSet<Text> partitionKeys = new TreeSet<>();
+    for (int split = 100000; split < 1000000; split += 100000) {
```
nit: make these `100_000` and `1_000_000` for readability

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:
```diff
@@ -835,7 +835,7 @@ void setSamplerConfiguration(String tableName, SamplerConfiguration samplerConfi
    * @see Summarizer
    */
-  SummaryRetriever getSummaries(String tableName) throws TableNotFoundException, AccumuloException, AccumuloSecurityException;
+  SummaryRetriever summarize(String tableName) throws TableNotFoundException, AccumuloException, AccumuloSecurityException;
```
I think `summarize` is a good improvement!

joshelser commented on core/src/main/java/org/apache/accumulo/core/client/summary/Summary.java:
```diff
@@ -78,27 +80,38 @@ public long getLarge() {
+     *         must be avoided, then compactions must be stopped. Compactions could be stopped on a cloned table to avoid this.
+     */
+    public long getDeleted() {
+      return deleted;
```
Is this onerous to compute? If this is rare and there is a workaround, I could see keeping this from public API. If it's not bad to compute/track, it's not an issue.

keith-turner replied:
It was not onerous to compute.  Special action needs to be taken on file not found anyway (do not want to retry).  In addition to ignoring the file, just need to report the info back.  Reporting back was not too bad as there were already other file stats.

I could not figure out to write a test for it.  I tested it manually.
